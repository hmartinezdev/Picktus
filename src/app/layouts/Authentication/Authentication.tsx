import Logo from '@components/Logo';
import colors from '@constants/colors';
import { borderRadius, boxShadow } from '@constants/styles';
import React, { Component } from 'react';
import { Transition, TransitionGroup } from 'react-transition-group';

export interface IAuthenticationPropsType {
  status: string;
  children: JSX.Element;
}

class Authentication extends Component<IAuthenticationPropsType> {
  public render(): React.ReactElement<Authentication> {
    const { status, children } = this.props;

    return (
      <div className={`container container--${status}`}>
        <div className={`logo`}>
          <Logo />
        </div>
        <TransitionGroup className="authentication__content" appear={true}>
          {React.Children.map(children, (child: JSX.Element) => (
            <Transition key={child.key || ''} timeout={300} mountOnEnter={true} unmountOnExit={true} appear={true}>
              {(contentStatus) => <div className={`contentContainer contentContainer--${contentStatus}`}>{child}</div>}
            </Transition>
          ))}
        </TransitionGroup>
        <style jsx>
          {`
            .container {
              width: 100%;
              height: 100vh;
              display: flex;
              align-items: center;
              flex-direction: column;
              padding: 2rem 0;
              box-sizing: border-box;
              overflow: hidden;
              opacity: 1;
              transition: opacity 300ms ease-out;
            }

            .container--exiting,
            .container--exited {
              position: absolute;
              z-index: 0;
            }

            .contentContainer {
              transition: opacity 300ms ease-out;
            }

            .contentContainer--entering {
              animation: opacity 300ms ease-in;
            }

            .contentContainer--entered {
              opacity: 1;
            }

            .contentContainer--exiting,
            .contentContainer--exited {
              position: absolute;
              opacity: 0;
            }

            .logo {
              background-color: ${colors.primary};
              box-shadow: ${boxShadow};
              display: flex;
              position: relative;
              border-radius: ${borderRadius};
              transition: opacity 300ms ease-out;
            }

            :global(.authentication__content) {
              height: 100%;
              width: 100%;
              display: flex;
              flex-direction: column;
              justify-content: flex-start;
              align-items: center;
              margin-top: 2.5rem;
            }

            @keyframes opacity {
              0% {
                opacity: 0;
              }

              100% {
                opacity: 1;
              }
            }
          `}
        </style>
      </div>
    );
  }
}

export default Authentication;
