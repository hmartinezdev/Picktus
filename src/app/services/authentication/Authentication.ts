import { session } from '@constants/cookies';
import * as firebase from 'firebase/app';
import 'firebase/auth';
import fetch from 'isomorphic-unfetch';
import Cookies from 'js-cookie';
import {
  delegatedMethods,
  FirebaseErrorCodes,
  ISigninMethodsMap,
  ISigninOptions,
  SigninMethods,
} from './Authentication.type';
import AuthenticationError from './AuthenticationError';
import { FirebaseError } from './AuthenticationError/AuthenticationError.type';

class Authentication {
  public async createUser(email: string, password: string) {
    return firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .catch((error: FirebaseError) => {
        throw this.handleFirebaseError(error, 'createUser');
      });
  }

  /**
   * Function handling the the social network signin
   *
   * @param method - method of authentication
   */
  private async delegatedAuthentication(method: delegatedMethods) {
    const providers: ISigninMethodsMap = {
      [SigninMethods.FACEBOOK]: new firebase.auth.FacebookAuthProvider(),
      [SigninMethods.GOOGLE]: new firebase.auth.GoogleAuthProvider(),
      [SigninMethods.TWITTER]: new firebase.auth.TwitterAuthProvider(),
      [SigninMethods.CLASSIC]: undefined,
    };

    const provider = providers[method];

    // Use the navigator default language
    firebase.auth().useDeviceLanguage();

    if (method === SigninMethods.FACEBOOK) {
      provider.setCustomParameters({
        display: 'popup',
      });
    }

    return firebase
      .auth()
      .signInWithPopup(provider)
      .catch((error) => {
        throw this.handleFirebaseError(error, 'delegatedAuthentication');
      });
  }

  /**
   * Function handling the whole signin process
   *
   * @param method - method of authentication, can be either classic or via social networks
   * @param options - options needed to signin with a specific method
   */
  public async signin(method: SigninMethods, options: ISigninOptions = {}) {
    let userCredentials: firebase.auth.UserCredential;
    if (method === SigninMethods.CLASSIC) {
      if (!options.email || !options.password) {
        throw new AuthenticationError(
          "Authentication::signin You can't use classic signin method without password or email passed as option",
          'A password and an email is necessary to signin'
        );
      }

      userCredentials = await firebase
        .auth()
        .signInWithEmailAndPassword(options.email, options.password)
        .catch((error) => {
          throw this.handleFirebaseError(error, 'signin');
        });
    } else {
      userCredentials = await this.delegatedAuthentication(method);
    }

    if (!userCredentials || !userCredentials.user) {
      throw new AuthenticationError(
        `Authentication::signin No user retrieved from firebase`,
        'We have a problem with the authentication server'
      );
    }

    const user = userCredentials.user;

    // Get the user's ID token as it is needed to exchange for a session cookie.
    user.getIdToken().then((idToken) => {
      // Session login endpoint is queried and the session cookie is set.
      // CSRF protection should be taken into account.
      // ...

      return fetch('/sessionlogin', {
        body: JSON.stringify({ idToken }),
        headers: {
          'Content-Type': 'application/json',
        },
        method: 'POST',
      });
    });

    return user;
  }

  public async disconnect() {
    await firebase
      .auth()
      .signOut()
      .catch((error) => {
        throw this.handleFirebaseError(error, 'signout');
      });

    // Remove the cookie set in session
    Cookies.remove(session);

    // reloading page to ensure the user is disconnected
    if (window && window.location) {
      window.location.reload();
    }
  }

  /**
   * Handle the generation of formated error
   *
   * @param error - error dispatched by firebase
   * @param func - function from which the error has been thrown
   */
  public handleFirebaseError(error: FirebaseError, func: string) {
    switch (error.code) {
      case FirebaseErrorCodes.TOO_MANY_REQUEST:
        return new AuthenticationError(
          `Authentication::${func} too many requests`,
          'You tried too many time to signin, please try again later'
        );
      case FirebaseErrorCodes.NETWOR_REQUEST_FAILED:
        return new AuthenticationError(
          `Authentication::${func} network error`,
          'You are currently in a slow network area, please try again later'
        );
      case FirebaseErrorCodes.INVALID_API_KEY:
        return new AuthenticationError(
          `Authentication::${func} invalid API provided`,
          'A client error has occured, please contact the consumer service'
        );
      case FirebaseErrorCodes.INVALID_EMAIL:
        return new AuthenticationError(
          `Authentication::${func} invalid email provided by the user`,
          'The email you provided is not a valid email address'
        );
      case FirebaseErrorCodes.USER_NOT_FOUND:
        return new AuthenticationError(
          `Authentication::${func} email was not associated with an account`,
          'The email you provided is not associated with an existing account'
        );
      case FirebaseErrorCodes.WRONG_PASSWORD:
        return new AuthenticationError(
          `Authentication::${func} password provided did not correspond with the email`,
          'The password and email combination is not associated with an existing account'
        );
      case FirebaseErrorCodes.ACCOUNT_ALREADY_EXIST_WITH_DIFFERENT_CREDENTIAL:
      case FirebaseErrorCodes.EMAIL_ALREADY_USED:
        return new AuthenticationError(
          `Authentication::${func} an account already exists with the user email`,
          'An account is already associated with this email'
        );
      case FirebaseErrorCodes.POPUP_BLOCKED:
        return new AuthenticationError(
          `Authentication::${func} popin blocked by navigator`,
          'The authentication popin has been blocked by your navigator'
        );
      case FirebaseErrorCodes.WEAK_PASSWORD:
        return new AuthenticationError(
          `Authentication::${func} popin blocked by navigator`,
          'The authentication popin has been blocked by your navigator'
        );
      case FirebaseErrorCodes.CANCELLED_POPUP:
        return new AuthenticationError(
          `Authentication::${func} popin cancelled because a new one has been opened`,
          'The previous authentication popup has been closed because you opened a new one'
        );
      default:
        return new AuthenticationError(
          `Authentication::${func} Unknown error: ${error.code}::${error.message}`,
          'An unknown error has occurred'
        );
    }
  }
}

export default new Authentication();
