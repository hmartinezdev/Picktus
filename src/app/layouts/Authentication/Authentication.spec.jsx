import { shallow } from 'enzyme';
import React from 'react';
import Authentication from './Authentication';

let wrapper;

const setup = () =>
  shallow(
    <Authentication>
      <div />
    </Authentication>
  );

describe('<Home />', () => {
  test('should render properly', () => {
    wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
