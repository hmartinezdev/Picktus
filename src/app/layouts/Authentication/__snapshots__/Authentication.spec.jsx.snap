// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render properly 1`] = `
<div
  className="container container--undefined"
>
  <div
    className="logo logo--undefined"
  >
    <Logo />
  </div>
  <div
    className="content"
  >
    <div
      className="contentContainer contentContainer--undefined"
    >
      <div />
    </div>
  </div>
  <style
    jsx={true}
  >
    
            .container {
              width: 100%;
              height: 100vh;
              display: flex;
              align-items: center;
              flex-direction: column;
              padding: 2rem 0;
              box-sizing: border-box;
              overflow: hidden;
              opacity: 1;
              transition: opacity 300ms ease-out;
            }

            .container--exiting,
            .container--exited {
              position: absolute;
              z-index: 0;
            }

            .contentContainer {
              transform: translateX(0%);
              transition: all 300ms ease-out;
            }

            .contentContainer--entering {
              animation: opacity 300ms ease-in;
            }

            .contentContainer--entered {
              opacity: 1;
            }

            .contentContainer--exiting,
            .contentContainer--exited {
              position: absolute;
              opacity: 0;
              transform: translateX(100%);
            }

            .logo {
              background-color: #2d3047;
              box-shadow: 11px 10px 34px -7px rgba(0, 0, 0, 0.49);
              display: flex;
              position: relative;
              border-radius: 3px;
              transition: opacity 300ms ease-out;
            }

            .logo--exiting {
              box-shadow: initial;
            }

            .content {
              height: 100%;
              width: 100%;
              display: flex;
              flex-direction: column;
              justify-content: flex-start;
              align-items: center;
              margin-top: 2.5rem;
            }

            @keyframes opacity {
              0% {
                opacity: 0;
                transform: translateX(-100%);
              }

              100% {
                opacity: 1;
                transform: translateX(0);
              }
            }
          
  </style>
</div>
`;
